src/Development/IDE/Core/Rules.hs

```diff
@@ -127,88 +131,102 @@ defineEarlyCutOffNoFile f = defineEarlyCutoff $ \k file -> do
+      cachedir <- liftIO getCurrentDirectory
+      let dir = cachedir </> "cached-deps" </> unitIdString uid
+          src = dir </> map (\x -> if x == '.' then '-' else x) (moduleNameString mod) <.> ext
+          ext = if boot then "hs-boot" else "hs"
+      hieTimestamp <- MaybeT $ either (\(_ :: IOException) -> Nothing) Just <$> (liftIO $ try $ getModificationTime hie_f)
+      srcTimestamp   <- either (\(_ :: IOException) -> Nothing) Just <$> (liftIO $ try $ getModificationTime src)
+      case srcTimestamp of
+        Just t | t >= hieTimestamp -> do
+          liftIO $ L.logInfo (logger ide) $ T.pack $ "Returning file " ++ show src
+          pure $ fromNormalizedUri $ filePathToUri' $ toNormalizedFilePath' src
+        _ -> do
+          hf <- readHieFileFromDisk hie_f
+          liftIO $ do
+            createDirectoryIfMissing True dir
+            removePathForcibly src
+            BS.writeFile src $ hie_hs_src hf
+            p <- getPermissions src
+            setPermissions src (p {writable = False})
+            writeChan dbchan $ \db -> HieDb.addSrcFile db hie_f src False
+            pure $ fromNormalizedUri $ filePathToUri' $ toNormalizedFilePath' src
+
 -- | Get all transitive file dependencies of a given module.
 -- Does not include the file itself.
 getDependencies :: NormalizedFilePath -> Action (Maybe [NormalizedFilePath])
 getDependencies file = fmap transitiveModuleDeps <$> use GetDependencies file
 
 -- | Try to get hover text for the name under point.
 getAtPoint :: NormalizedFilePath -> Position -> IdeAction (Maybe (Maybe Range, [T.Text]))
-getAtPoint file pos = fmap join $ runMaybeT $ do
+getAtPoint file pos = runMaybeT $ do
   ide <- ask
   opts <- liftIO $ getIdeOptionsIO ide
 
-  (hieAst -> hf, mapping) <- useE GetHieAst file
+  (hf, mapping) <- useE GetHieAst file
   dkMap <- lift $ maybe (DKMap mempty mempty) fst <$> (runMaybeT $ useE GetDocMap file)
 
   !pos' <- MaybeT (return $ fromCurrentPosition mapping pos)
-  return $ AtPoint.atPoint opts hf dkMap pos'
+  MaybeT $ pure $ fmap (first (toCurrentRange mapping =<<)) $ AtPoint.atPoint opts hf dkMap pos'
+
+toCurrentLocations :: PositionMapping -> [Location] -> [Location]
+toCurrentLocations mapping = mapMaybe go
+  where
+    go (Location uri range) = Location uri <$> toCurrentRange mapping range
 
 -- | Goto Definition.
-getDefinition :: NormalizedFilePath -> Position -> IdeAction (Maybe Location)
+getDefinition :: NormalizedFilePath -> Position -> IdeAction (Maybe [Location])
 getDefinition file pos = runMaybeT $ do
     ide <- ask
     opts <- liftIO $ getIdeOptionsIO ide
-    (HAR _ hf _ , mapping) <- useE GetHieAst file
+    (HAR _ hf _ _, mapping) <- useE GetHieAst file
     (ImportMap imports, _) <- useE GetImportMap file
     !pos' <- MaybeT (return $ fromCurrentPosition mapping pos)
-    AtPoint.gotoDefinition (getHieFile ide file) opts imports hf pos'
+    hiedb <- lift $ asks hiedb
+    hiedbChan <- lift $ asks hiedbWriter
```

Comment:
What sort of guarantees are given about the hiedb contents? Is it up-to-date?

I'm surprised it isn't managed by a Shake rule

Reply:
No, the invariant is that it is up to date when `pendingIndexes` is empty and we've received no more edits.

Reply:
Can we document this somewhere?
What depends on this invariant and how is it enforced?

Reply:
OK. It's enforced by us queuing up an index operation every time a `.hie` file is generated - which is every time the `GetModIface` rule is recomputed. Also, when reading a `.hi` file from disk, we check the database to ensure that the the corresponding `.hie` file has been indexed, and if not, we re-index it. This maintains database consistency if ghcide is killed before all the indexing completes.

Nothing "depends" on the invariant as such, since we are after eventual consistency - the results may be out of date at any given time.

Reply:
It would be good to write this down in a comment somewhere

Do we make any effort to avoid redundant re-indexings ?

Reply:
> Do we make any effort to avoid redundant re-indexings ?

Yes, we only re-index only if one of the following conditions hold:

-  The `.hi` file had to be regenerated
- The `.hs` file is newer than the `.hie` file.